Replace java-jwt with nimbus-jose-jwt

Arnei opened the pull request:
Fix #6093.

The goal of this commit is to introduce a jwt library capable of handling EdDSA signed jwts. java-jwt does not do this and it is currently unclear when or even if it ever will be able to. nimbus-jose-jwt has been chosen because it:
- supports EdDSA (through google-crypto-tink)
- has a long-standing history
- is actively developed
- is easy to use

### How to test this patch

Ideally you already have a jwt setup, so you can test if it still works. Be aware that the configuration has changed slightly (no functional changes, but one variable has been renamed and explicit string transformations are now unnecessary).

### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://docs.opencast.org/develop/developer/#participate/development-process/#automatically-closing-issues-when-a-pr-is-merged) if one exists
* [ ] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)
* [ ] explain why it needs to be merged into the legacy branch, if it is targeting the legacy branch

Arnei committed:
Replace java-jwt with nimbus-jose-jwt

Fix #6093.

The goal of this commit is to introduce a jwt library capable of
handling EdDSA signed jwts. java-jwt does not do this and it is
currently unclear when or even if it will be able to.
nimbus-jose-jwt has been chosen because it:
- supports EdDSA (through google-crypto-tink)
- has a long-standing history
- is actively developed
- is easy to use

LukasKalbertodt commented on etc/security/mh_default_org.xml:
```diff
@@ -780,7 +780,18 @@
+    <value>ES384</value>
+    <value>ES512</value>
+    <value>EdDSA</value>
+    <value>Ed25519</value>
```
Why are `EdDSA` and `Ed25519` valid values here? `Ed25519` is a subset of `EdDSA`, or rather one specific curve. `EdDSA` just describes all signature schemes with twisted edwards curves, which also includes `Ed448` (another specific curve, hardly supported). I do get that it might make sense to explicitly want to allow only a single specific curve or to generally allow a family, but given that only `EdDSA` is a valid `alg` value, this might confuse people. 

I would probably just use `EdDSA` and not give users the option to only specify a single curve. This configuration is weird anyway, as there is little benefit to restricting algorithms like this, as far as I know.

Arnei replied:
Both are valid values because I did not think about this too much. I simply saw that the nimbus library offered both values as alg constant, so I added them. I'll change it to just `EdDSA` then.

LukasKalbertodt commented on etc/security/mh_default_org.xml:
```diff
@@ -765,13 +765,13 @@
     <property name="userReferenceProvider" ref="userReferenceProvider" />
     <property name="jwksUrl" value="https://auth.example.org/.well-known/jwks.json" />
-    <property name="jwksCacheExpiresIn" value="1440" />
+    <property name="jwksTimeToLive" value="3600000" />
```
Why the unrelated change of default duration? 

And as a different issue: this seems like a breaking change that's not really required. We could keep the old name and unit (seconds) in this file to make updating easier. Probably not super important because there are other breaking changes that we cannot avoid, but still: fewer breaking changes are nice when there is no good reason to make htem.

Arnei replied:
I changed this because
- Internally this was previously minutes, but is now milliseconds.
- An hour (3600000 ms) was recommended by the nimbus-jose-jwt docs as a reasonable default.

I have no hard opinion on this however. If you think it is better, I'll change it back to the old value (a day) and parse it internally from minutes to milliseconds.

LukasKalbertodt commented on modules/security-jwt/src/main/java/org/opencastproject/security/jwt/JWTVerifier.java:
```diff
@@ -21,55 +21,73 @@
+      ECKey ecKey = jwkSet.get(0).toECKey();
+      return verify(jwt, claimConstraints, new ECDSAVerifier(ecKey));
+    } else if (alg.equals(JWSAlgorithm.EdDSA) || alg.equals(JWSAlgorithm.Ed25519)) {
+      JWK jwk = jwkSet.get(0).toPublicJWK();
```
All this code only tries the first key of the JWKS? Am I reading this correctly? Why is that? Was that the previous behavior as well?

mtneug replied:
@joneug do you know what the previous behavior was? If we only supported the first entry, IMO this would be a bug and we should support JWKS with multiple entries.

joneug replied:
In the current implementation, all keys are fetched from the provider.

https://github.com/opencast/opencast/blob/8046ae9b7fd3a67bc9581a308e3710c95cebcabd/modules/security-jwt/src/main/java/org/opencastproject/security/jwt/JWTVerifier.java#L63-L73

The verification logic then considers all of them:

https://github.com/opencast/opencast/blob/8046ae9b7fd3a67bc9581a308e3710c95cebcabd/modules/security-jwt/src/main/java/org/opencastproject/security/jwt/JWTVerifier.java#L126-L154

Arnei replied:
Thanks for pointing that out, should be fixed now.

LukasKalbertodt reviewed (commented):
I haven't really looked at the code, but I tested this. And it works with Tobira! The old ES256 key still works, but Ed25519 now works too.

Arnei committed:
Use all keys in the JWKS again

Was only using the first key of the set, is now using all keys.
This is consistent with earlier behaviour when using java-jwt.

Arnei committed:
Only offer EdDSA in mh_default_org.xml

Removes Ed25519, as EdDSA encompasses it.
